File: leetcode51/NQueens.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class NQueens {

    public static void main(String[] args) {
        NQueens nq = new NQueens();
        int n = 14;
        long start = System.currentTimeMillis();
        List<List<String>> ans = nq.solveNQueens(n);
        long span = System.currentTimeMillis() - start;
        System.out.printf("answer for %d-queens:%n", n);
        for (int i = 0; i < ans.size(); i++) {
            List<String> subAns = ans.get(i);
            System.out.printf("solution %d:%n", i+1);
            for (String s : subAns) {
                System.out.printf("%s%n", s);
            }
            System.out.println("");
        }
        System.out.printf("time taken: %d%n", span);
    }

    /**
     * Runtime: 1 ms, faster than 99.97% of Java online submissions for N-Queens.
     * Memory Usage: 39.8 MB, less than 69.22% of Java online submissions for N-Queens.
     */
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        int[] subResult = new int[n];
        boolean[] rowAttack = new boolean[n];
        boolean[] diag1Attack = new boolean[2*n-1];
        boolean[] diag2Attack = new boolean[2*n-1];
        placeQueen(n, 0, subResult, result,
                   rowAttack, diag1Attack, diag2Attack);
        return result;
    }

    private void placeQueen(int n, int c,
        int[] subResult, List<List<String>> result,
        boolean[] rowAttack, boolean[] diag1Attack, boolean[] diag2Attack) {
        for (int r = 0; r < n; r++) {
            if (rowAttack[r] || diag1Attack[r+c] || diag2Attack[r-c+n-1]) continue;
            subResult[c] = r;
            int nc = c + 1;
            if (nc == n) {
                List<String> list = new ArrayList<>(n);
                for (int i = 0; i < n; i++) {
                    char[] row = new char[n];
                    for (int j = 0; j < n; j++) {
                        if (subResult[j] == i) {
                            row[j] = 'Q';
                        } else {
                            row[j] = '.';
                        }
                    }
                    list.add(String.valueOf(row));
                }
                result.add(list);
                return;
            }
            rowAttack[r] = true;
            diag1Attack[r+c] = true;
            diag2Attack[r-c+n-1] = true;
            placeQueen(n, nc, subResult, result, rowAttack, diag1Attack, diag2Attack);
            rowAttack[r] = false;
            diag1Attack[r+c] = false;
            diag2Attack[r-c+n-1] = false;
        }
    }
}

File: leetcode51/n_queens.rs
use std::time::Instant;

pub fn main() {
    let n = 14;
    let now = Instant::now();
    let ans = solve_n_queens(n);
    let time_taken_ms = now.elapsed().as_millis();
    println!("{}-queens problem, answer:", n);
    for (i, solution) in ans.iter().enumerate() {
        println!("solution {}", i);
        for row in solution.iter() {
            println!("{:?}", row);
        }
    }
    println!("time taken ms: {}", time_taken_ms);
}

///
/// Runtime: 0 ms, faster than 100.00% of Rust online submissions for N-Queens.
/// Memory Usage: 2.3 MB, less than 77.27% of Rust online submissions for N-Queens.
pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
    let n1 = n as usize;
    let mut sub_ans = vec![0usize; n1];
    let mut ans = Vec::new();
    let mut c_atk = vec![false; n1];
    let mut d1_atk = vec![false; 2*n1-1];
    let mut d2_atk = vec![false; 2*n1-1];
    place_queen(n1, 0, &mut sub_ans, &mut ans, &mut c_atk,
    &mut d1_atk, &mut d2_atk);
    ans
}

fn place_queen(n: usize, r: usize, sub_ans: &mut Vec<usize>,
        ans: &mut Vec<Vec<String>>, c_atk: &mut Vec<bool>,
        d1_atk: &mut Vec<bool>, d2_atk: &mut Vec<bool>) {
    for c in 0..n {
        if !c_atk[c] && !d1_atk[r+c] && !d2_atk[n+r-c-1] {
            sub_ans[r] = c;
            if r+1 == n {
                let mut rows = Vec::with_capacity(n);
                for i in 0..n {
                    let mut chars = vec!['.'; n];
                    chars[sub_ans[i]] = 'Q';
                    let row: String = chars.iter().collect();
                    rows.push(row);
                }
                ans.push(rows);
                return;
            }
            c_atk[c] = true;
            d1_atk[r+c] = true;
            d2_atk[n+r-c-1] = true;
            place_queen(n, r+1, sub_ans, ans, c_atk, d1_atk, d2_atk);
            c_atk[c] = false;
            d1_atk[r+c] = false;
            d2_atk[n+r-c-1] = false;
        }
    }
}
